[AArch64] Extend efficient lowering of experimental.cttz.elts

hazzlim opened the pull request:
This patch extends support for more efficient lowering of the experimental.cttz.elts intrinsic to fixed-width vector types, by first creating an SVE predicate register mask from the fixed-width vector.

hazzlim committed:
[AArch64] Extend efficient lowering of experimental.cttz.elts

This patch extends support for more efficient lowering of the
experimental.cttz.elts intrinsic to fixed-width vector types, by first
creating an SVE predicate register mask from the fixed-width vector.

hazzlim commented:
Requesting review from @david-arm @paulwalker-arm @davemgreen

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5838,9 +5840,21 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+    SDValue CttzOp = Op.getOperand(1);
+    EVT VT = CttzOp.getValueType();
+
+    if (!VT.isScalableVector()) {
```
It's better to check for the affirmative (in this case `isFixedLengthVector()`) when protecting code.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5838,9 +5840,21 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+
+    if (!VT.isScalableVector()) {
+      // Retrieve original fixed-width vector from ISD::TRUNCATE Node.
+      assert(CttzOp.getOpcode() == ISD::TRUNCATE && "Expected ISD::TRUNCATE!");
```
This is not a safe assumption to make.  Also, "masks" have a special format as dictated by `getBooleanContents` which `convertFixedMaskToScalableVector` is allowed to depend on.

Putting this together I think you should emit a suitable compare against zero to construct a mask that can be passed to `convertFixedMaskToScalableVector`.  I'd expect DAG combine to do a good job of removing this compare for the truncate scenario your expecting.

paulwalker-arm replied:
Sorry, I hit the wrong button and thus posted my review comments early whilst still investigating the code.

I now understand where the truncate comes from but as before, this is something that cannot be relied upon.  For AArch64 `getBooleanContents` requires fixed length masks to be all zeros or all ones.  Given the change to `shouldExpandCttzElements` we can be sure the operand type will be an i1 vector (which you can and should assert for) and thus I think you can just sign extend to the relevant type (see `getTypeToTransformTo`) and pass this to `convertFixedMaskToScalableVector`.  As before, I'd expect this sign extend will be optimised away.

hazzlim committed:
Properly sign extend fixed-width i1 vectors to legal type

Note: this produces a sub-optimal lowering, as the sign_extend is not
optimised away. This is left to a follow-up patch.

hazzlim commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5838,9 +5840,21 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
+
+    if (!VT.isScalableVector()) {
+      // Retrieve original fixed-width vector from ISD::TRUNCATE Node.
+      assert(CttzOp.getOpcode() == ISD::TRUNCATE && "Expected ISD::TRUNCATE!");
```
Thanks for pointing this out - I have updated so that we sign extend to the correct type before passing this (now valid) mask to `convertFixedMaskToScalableVector`.

We don't actually manage to optimise away the sign_extend, at present - I have updated the regression tests to reflect this as it seems to make sense to fix this in a separate, follow up patch?

paulwalker-arm replied:
Agreed, the DAG combine should be a separate PR.
